lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,207 @@
+// NewKubeAppsFetcher creates new Kubernetes app fetcher
+func NewKubeAppsFetcher(cfg KubeAppsFetcherConfig) (common.Fetcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &kubeAppFetcher{
+		KubeAppsFetcherConfig: cfg,
+	}, nil
+}
+
+func isInternalKubeService(s v1.Service) bool {
+	const kubernetesDefaultServiceName = "kubernetes"
+	return (s.GetNamespace() == metav1.NamespaceDefault && s.GetName() == kubernetesDefaultServiceName) ||
+		s.GetNamespace() == metav1.NamespaceSystem ||
+		s.GetNamespace() == metav1.NamespacePublic
+}
+
+func (f *kubeAppFetcher) getServices(ctx context.Context) ([]v1.Service, error) {
+	var result []v1.Service
+	nextToken := ""
+	namespaceFilter := func(ns string) bool {
+		return slices.Contains(f.Namespaces, types.Wildcard) || slices.Contains(f.Namespaces, ns)
+	}
+	for {
+		// Get all services in the cluster
+		// We need to do this in a loop because the API only returns 500 items at a time
+		// and we need to paginate through the results.
+		kubeServices, err := f.KubernetesClient.CoreV1().Services(v1.NamespaceAll).List(
+			ctx,
+			metav1.ListOptions{
+				Continue: nextToken,
+			})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, s := range kubeServices.Items {
+			if !namespaceFilter(s.GetNamespace()) {
+				// Namespace is not in the list of namespaces to fetch
+				continue
+			}
+			match, _, err := services.MatchLabels(f.FilterLabels, s.Labels)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			} else if match {
+				result = append(result, s)
+			} else {
+				f.Log.WithField("service_name", s.Name).Debug("Service doesn't match labels.")
+			}
+		}
+		nextToken = kubeServices.Continue
+		if nextToken == "" {
+			break
+		}
+	}
+	return result, nil
+}
+
+// Get fetches Kubernetes apps from the cluster
+func (f *kubeAppFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	kubeServices, err := f.getServices(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Converting service to apps can involve performing a HTTP ping to the service ports to determine protocol.
+	// Both services and ports inside services are processed in parallel to minimize time.
+	// We also set limit to prevent potential spike load on a cluster in case there are a lot of services.
+	g, _ := errgroup.WithContext(ctx)
+	g.SetLimit(10)
+
+	// Convert services to resources
+	var (
+		appsMu sync.Mutex
+		apps   types.Apps
+	)
+	for _, service := range kubeServices {
+		service := service
+
+		g.Go(func() error {
+			// Skip kubernetes own internal services
+			if isInternalKubeService(service) {
+				return nil
+			}
+
+			// Skip service if it has type annotation and it's not 'app'
+			if v, ok := service.GetAnnotations()[types.DiscoveryTypeLabel]; ok && v != services.KubernetesMatchersApp {
+				return nil
+			}
+
+			serviceApps, err := services.NewApplicationsFromKubeService(service, f.ClusterName, f.protocolChecker)
```

Comment:
this already uses parallelism inside so I think my comment in the other PR makes more sense otherwise we can have 10*5 parallel runs